### PR TITLE
feat(bigtable): start converting bigtable to use `g::c::Options`

### PR DESCRIPTION
**Pull request by dbolduc**
`ClientOptions` can now be constructed with `Options`. Users are recommended to use this constructor over the one with `grpc::ChannelCredentials`. `ClientOptions` also uses `Options` internally.

Part of the work for #6307 

`channel_arguments_` still need to be incorporated into `Options opts_`, but I thought that was involved enough to warrant its own PR.

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6949)

### CODE REVIEW

**Review by coryan — CHANGES_REQUESTED** [review]
> Reviewed 8 of 8 files at r1.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6949)* status: all files reviewed, 5 unresolved discussions (waiting on @dbolduc)

___
*[google/cloud/bigtable/client_options.cc, line 91 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6949#-Me21L_f0sZQ6DM9Zj6S:-Me21L_f0sZQ6DM9Zj6T:b8i8bsv) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/4ea56eb141558550f0f85badf58464baab8bf548/google/cloud/bigtable/client_options.cc#L91)):*
> ```cpp
> ClientOptions& ClientOptions::set_connection_pool_size(std::size_t size) {
>   if (size == 0) {
>     opts_.set<ConnectionPoolSizeOption>(internal::DefaultConnectionPoolSize());
> ```

nit: I think we would use a `size != 0 ? size : internal::DefaultConnectionPoolSize()` these days.
___
*[google/cloud/bigtable/options.h, line 52 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6949#-Me2-2TD8QWz6G-MKx3g:-Me2-2TD8QWz6G-MKx3h:b4sgbtt) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/4ea56eb141558550f0f85badf58464baab8bf548/google/cloud/bigtable/options.h#L52)):*
> ```objc
> 
> /// The connection pool size.
> struct ConnectionPoolSizeOption {
> ```

There is an existing `GrpcNumChannels` option with the same semantics, why not use it?

https://github.com/googleapis/google-cloud-cpp/blob/6bc7fd8b1a7c7ce28e359b959b5ee37dd1d52b38/google/cloud/grpc_options.h#L44-L46
___
*[google/cloud/bigtable/options.h, line 57 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6949#-Me2-pO60WmqTX217Jbq:-Me2-7sdDGuE0sEsxrZc:b16v4ea) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/4ea56eb141558550f0f85badf58464baab8bf548/google/cloud/bigtable/options.h#L57)):*
> ```objc
> 
> /// The connection pool name.
> struct ConnectionPoolNameOption {
> ```

Optional: this seems like it would be generally useful, maybe refactor to `google/cloud/grpc_options.h`?  
___
*[google/cloud/bigtable/options.h, line 98 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6949#-Me204hw3xgsIliV6588:-Me204hw3xgsIliV6589:b1atg1q) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/4ea56eb141558550f0f85badf58464baab8bf548/google/cloud/bigtable/options.h#L98)):*
> ```objc
>  * @note this value is not used if background threads are disabled.
>  */
> struct BackgroundThreadPoolSizeOption {
> ```

This seems like it would be generally useful, can we refactor too?
___
*[google/cloud/bigtable/internal/defaults.cc, line 62 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6949#-Me20SsL8yYZF5UzXKBR:-Me20SsL8yYZF5UzXKBS:b-sepejh) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/4ea56eb141558550f0f85badf58464baab8bf548/google/cloud/bigtable/internal/defaults.cc#L62)):*
> ```cpp
> 
> Options DefaultOptions(Options opts) {
>   auto emulator = google::cloud::internal::GetEnv("BIGTABLE_EMULATOR_HOST");
> ```

It might be possible to use 

https://github.com/googleapis/google-cloud-cpp/blob/6bc7fd8b1a7c7ce28e359b959b5ee37dd1d52b38/google/cloud/options.h#L272-L276

To simplify this code... it would become something like:

```cc
return google::cloud::internal::MergeOptions(std::move(opts), Options{}.set<..>(default_1).set<...>(default_2)...);
```

**Comment by dbolduc** [comment]
FYI, I am going to work on #6955. When it is done, I will return to this PR and address the rest of the review comments.

In the mean time, I am marking the PR as a draft.

**Review by dbolduc — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6949)* status: 2 of 22 files reviewed, 4 unresolved discussions (waiting on @coryan)

___
*[google/cloud/bigtable/client_options.cc, line 91 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6949#-Me21L_f0sZQ6DM9Zj6S:-MeCUNE374dG2_kMNlM8:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/4ea56eb141558550f0f85badf58464baab8bf548/google/cloud/bigtable/client_options.cc#L91)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

nit: I think we would use a `size != 0 ? size : internal::DefaultConnectionPoolSize()` these days.
</blockquote></details>

Done.
___
*[google/cloud/bigtable/options.h, line 52 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6949#-Me2-2TD8QWz6G-MKx3g:-MeCVnxX7sPXMfmRwiaq:b-x5e4ip) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/4ea56eb141558550f0f85badf58464baab8bf548/google/cloud/bigtable/options.h#L52)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

There is an existing `GrpcNumChannels` option with the same semantics, why not use it?

https://github.com/googleapis/google-cloud-cpp/blob/6bc7fd8b1a7c7ce28e359b959b5ee37dd1d52b38/google/cloud/grpc_options.h#L44-L46

</blockquote></details>

I did not realize they were the same. Done....

But shouldn't it be `using Type = std::size_t`? Is a negative number for this option supposed to mean something? Do we need to handle negative values?
___
*[google/cloud/bigtable/options.h, line 57 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6949#-Me2-pO60WmqTX217Jbq:-MeCXQikAVUk78WXUrPY:b-qnofq0) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/4ea56eb141558550f0f85badf58464baab8bf548/google/cloud/bigtable/options.h#L57)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Optional: this seems like it would be generally useful, maybe refactor to `google/cloud/grpc_options.h`?  
</blockquote></details>

TL;DR: I am going to revert the `connection_pool_name_` change and deal with it later

So upon further inspection, the connection pool name is really just a gRPC channel (string) argument in disguise. A [comment](https://github.com/googleapis/google-cloud-cpp/blob/690602fa9cbe771052aee16495284c014bd27e02/google/cloud/connection_options.h#L269-L276) in our `ConnectionOptions` says there is a more official way to do this using the `GRPC_ARG_CHANNEL_POOL_DOMAIN` [macro](https://grpc.github.io/grpc/core/group__grpc__arg__keys.html#ga126793d7d941d79b69b1446fc17989c9). [Here](https://github.com/googleapis/google-cloud-cpp/blob/690602fa9cbe771052aee16495284c014bd27e02/google/cloud/bigtable/internal/common_client.h#L211-L225) is how we currently use the connection pool name in Bigtable.

So because I said I would deal with `grpc::ChannelArguments` in the next PR, I am going to revert the `ConnectionPoolNameOption` changes for this PR.

Also, now that I'm down the rabbit hole, I am noticing that we use the `GrpcNumChannelsOption` to determine how many channels there are. In bigtable, pubsub, spanner, storage we [loop over the number of channels](https://github.com/googleapis/google-cloud-cpp/blob/690602fa9cbe771052aee16495284c014bd27e02/google/cloud/pubsub/publisher_connection.cc#L75-L81) and call something like [this](https://github.com/googleapis/google-cloud-cpp/blob/690602fa9cbe771052aee16495284c014bd27e02/google/cloud/pubsub/internal/create_channel.cc#L30). But we do no such thing in our generated code. Is this because we only ever want to make a single stub that uses a single channel for generated libraries? If so, then I think it is misleading to [set the value](https://github.com/googleapis/google-cloud-cpp/blob/690602fa9cbe771052aee16495284c014bd27e02/generator/internal/option_defaults_generator.cc#L105-L107) of `GrpcNumChannelsOption`.
___
*[google/cloud/bigtable/options.h, line 98 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6949#-Me204hw3xgsIliV6588:-MeCELsiE98NVaIx41lg:b5b7n8k) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/4ea56eb141558550f0f85badf58464baab8bf548/google/cloud/bigtable/options.h#L98)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

This seems like it would be generally useful, can we refactor too?
</blockquote></details>

Done. See #6962 
___
*[google/cloud/bigtable/internal/defaults.cc, line 62 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6949#-Me20SsL8yYZF5UzXKBR:-MeCyi4Z2iffdRHTneV1:b-ommsqy) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/4ea56eb141558550f0f85badf58464baab8bf548/google/cloud/bigtable/internal/defaults.cc#L62)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

It might be possible to use 

https://github.com/googleapis/google-cloud-cpp/blob/6bc7fd8b1a7c7ce28e359b959b5ee37dd1d52b38/google/cloud/options.h#L272-L276

To simplify this code... it would become something like:

```cc
return google::cloud::internal::MergeOptions(std::move(opts), Options{}.set<..>(default_1).set<...>(default_2)...);
```
 
</blockquote></details>

It would definitely be possible. I would push back on changing this now, because I am a fan of consistency and I think this is the most common way that we default our `Options`. (e.g. [spanner](https://github.com/googleapis/google-cloud-cpp/blob/690602fa9cbe771052aee16495284c014bd27e02/google/cloud/spanner/internal/defaults.cc) and our [generated libraries](https://github.com/googleapis/google-cloud-cpp/blob/690602fa9cbe771052aee16495284c014bd27e02/generator/internal/option_defaults_generator.cc)).

"Consistency is not a good reason to perpetuate existing bad design choices." - jgm

So I think a compromise would be to leave `bigtable/internal/defaults.cc` alone and open a new issue to refactor all such instances of defaulting `Options` at one time.

**Review by dbolduc — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6949)* status: 2 of 22 files reviewed, 4 unresolved discussions (waiting on @coryan)

___
*[google/cloud/bigtable/internal/defaults.cc, line 62 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6949#-Me20SsL8yYZF5UzXKBR:-MeDwcP-EVBXteNTTL_5:b256v4d) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/4ea56eb141558550f0f85badf58464baab8bf548/google/cloud/bigtable/internal/defaults.cc#L62)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

It would definitely be possible. I would push back on changing this now, because I am a fan of consistency and I think this is the most common way that we default our `Options`. (e.g. [spanner](https://github.com/googleapis/google-cloud-cpp/blob/690602fa9cbe771052aee16495284c014bd27e02/google/cloud/spanner/internal/defaults.cc) and our [generated libraries](https://github.com/googleapis/google-cloud-cpp/blob/690602fa9cbe771052aee16495284c014bd27e02/generator/internal/option_defaults_generator.cc)).

"Consistency is not a good reason to perpetuate existing bad design choices." - jgm

So I think a compromise would be to leave `bigtable/internal/defaults.cc` alone and open a new issue to refactor all such instances of defaulting `Options` at one time.
</blockquote></details>

opened #6965

**Review by coryan — CHANGES_REQUESTED** [review]
> Reviewed 20 of 20 files at r2.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6949)* status: all files reviewed, 3 unresolved discussions (waiting on @dbolduc)

___
*[google/cloud/bigtable/client_options.h, line 449 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6949#-MeFpmUnEnz9ZupMbdvT:-MeFpmUo4cRYA3_BgJSa:bllc5k7) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/4ea56eb141558550f0f85badf58464baab8bf548/google/cloud/bigtable/client_options.h#L449)):*
> ```objc
>       google::cloud::CompletionQueue const& cq);
> 
>   using BackgroundThreadsFactory =
> ```

This is an avoidable breaking change, consider:

```cc
/**
  * Backwards compatibility alias
  *
  * @deprecated Consider using `google::cloud::BackgroundThreadsFactory` directly
  */
using BackgroundThreadsFactory = ::google::cloud::BackgroundThreadsFactory;
```

We may be able to use a `[[deprecated]]` annotation, but I am not sure how those work for aliases, so maybe test those and consider adding them in a future PR.
___
*[google/cloud/bigtable/options.h, line 52 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6949#-Me2-2TD8QWz6G-MKx3g:-MeFrEzk1a7lnLxwvQWj:b-c3asbr) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/4ea56eb141558550f0f85badf58464baab8bf548/google/cloud/bigtable/options.h#L52)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

I did not realize they were the same. Done....

But shouldn't it be `using Type = std::size_t`? Is a negative number for this option supposed to mean something? Do we need to handle negative values?
</blockquote></details>

Changing the type to `std::size_t` seems like a separate discussion. I am not sure what we do with out of range option values.  Ignore them and use the default?  We cannot return errors where the `Options` are used.  Can you open a bug?
___
*[google/cloud/bigtable/options.h, line 71 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6949#-MeFqlVS8xcso3MxPtD9:-MeFqlVT1yw9yYA-p2RK:b-kryi4c) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/4d55cad5f2619692a01b394b130342fd5d996afe/google/cloud/bigtable/options.h#L71)):*
> ```objc
> };
> 
> /*
> ```

I think this should be a Doxygen comment so `/**` to start the block.
___
*[google/cloud/bigtable/options.h, line 79 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6949#-MeFqrLc6iVQ3-UskSoo:-MeFqrLd8sVcqjvVv5fr:b-ni0ay0) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/4d55cad5f2619692a01b394b130342fd5d996afe/google/cloud/bigtable/options.h#L79)):*
> ```objc
> };
> 
> /*
> ```

Same here about Doxygen.
___
*[google/cloud/bigtable/internal/defaults.cc, line 62 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6949#-Me20SsL8yYZF5UzXKBR:-MeFr-iq9Nzjyt6chtfv:br8m3fp) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/4ea56eb141558550f0f85badf58464baab8bf548/google/cloud/bigtable/internal/defaults.cc#L62)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

opened #6965 
</blockquote></details>

a followup bug SGTM.

**Review by dbolduc — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6949)* status: 20 of 22 files reviewed, 3 unresolved discussions (waiting on @coryan)

___
*[google/cloud/bigtable/client_options.h, line 449 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6949#-MeFpmUnEnz9ZupMbdvT:-MeGBg029DxG7BHlH636:b-vl28qx) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/4ea56eb141558550f0f85badf58464baab8bf548/google/cloud/bigtable/client_options.h#L449)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

This is an avoidable breaking change, consider:

```cc
/**
  * Backwards compatibility alias
  *
  * @deprecated Consider using `google::cloud::BackgroundThreadsFactory` directly
  */
using BackgroundThreadsFactory = ::google::cloud::BackgroundThreadsFactory;
```

We may be able to use a `[[deprecated]]` annotation, but I am not sure how those work for aliases, so maybe test those and consider adding them in a future PR.
</blockquote></details>

Yeah....... you are obviously right. 

You'd think I would have realized this was a breaking change when I had to fix all of the broken instances of `ClientOptions::BackgroundThreadsFactory` in our code..........

Done.
___
*[google/cloud/bigtable/options.h, line 52 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6949#-Me2-2TD8QWz6G-MKx3g:-MeGGiSD6ogo7jW7cQz2:b-8o45m4) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/4ea56eb141558550f0f85badf58464baab8bf548/google/cloud/bigtable/options.h#L52)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Changing the type to `std::size_t` seems like a separate discussion. I am not sure what we do with out of range option values.  Ignore them and use the default?  We cannot return errors where the `Options` are used.  Can you open a bug?
</blockquote></details>

Opened #6966
___
*[google/cloud/bigtable/options.h, line 71 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6949#-MeFqlVS8xcso3MxPtD9:-MeGHQiA6BocB7XZ1NwE:bt2vr1b) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/4d55cad5f2619692a01b394b130342fd5d996afe/google/cloud/bigtable/options.h#L71)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

I think this should be a Doxygen comment so `/**` to start the block.
</blockquote></details>

Fixed.
___
*[google/cloud/bigtable/options.h, line 79 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6949#-MeFqrLc6iVQ3-UskSoo:-MeGHhFzBH4L1XXHGFZN:bt2vr1b) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/4d55cad5f2619692a01b394b130342fd5d996afe/google/cloud/bigtable/options.h#L79)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Same here about Doxygen.
</blockquote></details>

Fixed.

**Review by coryan — APPROVED** [review]
:lgtm:

> Reviewed 1 of 2 files at r3, 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6949#-:-MeG_nnaCO6HkHpFJjpo:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @dbolduc)
